### DISCUSSION TITLE
Sync function

### DISCUSSION THREAD

**Discussion by adrian141b** [Rapptz/discord.py#8227] [General]
Hello,
I have a question about the discord bot function, Can i create the sync function to get messagess from discord? Async function create some troubles on my app.

**Comment by GoogolGenius** [comment]
What issues are you experiencing? Any traceback and minimal reproducible code you can provide?

**Reply by adrian141b** [reply]
wants to run the bot only once (something like the download all messages until now button) after downloading the messages, we need to be able to do it again, the asynchronous function does not allow this to happen

**Comment by trevorflahardy** [comment]
## Overview
Discord.py is an async library, meaning that what you're trying to achieve isn't going to yield a great solution. I'm going to show you two methods you can use and explain why both of them aren't great. Afterwards, I'm going to give you some questions to answer for yourself to you can hopefully find a solution that works for you.

### Async Approach
We can use [`asyncio`](https://docs.python.org/3/library/asyncio.html#module-asyncio)'s utilities to run coroutines in synchronous contexts. Let's take a look at how we could achieve this.

```python
import asyncio
import discord
from discord.ext import commands

bot = commands.Bot(command_prefix='!', intents=discord.Intents.default())
bot.http.token = 'my_bot_token' # Used for Authentication

# asyncio.run will only work if there's no current event loop set. If there is though, you may want to look into 
# asyncio.create_task.
channel: discord.TextChannel = asyncio.run(bot.fetch_channel(MY_CHANNEL_ID))
message: discord.Message = asyncio.run(channel.fetch_message(MY_MESSAGE_ID))
print(message.content)
```

### Using Requests
Nothing limits you from using a 3rd party library like [`requests`](https://pypi.org/project/requests/) to authenticate yourself into the API and make the requests manually. Let's look at how we could use this.

- [Get Channel Message](https://discord.com/developers/docs/resources/channel#get-channel-message): `/channels/channel.id/messages/message.id`

```python
import requests

BOT_TOKEN = 'my_bot_token' # Your bot's token, used for authentication
MY_CHANNEL_ID = 0 # The channel ID of the channel you want to fetch the message from.
MY_MESSAGE_ID = 0 # The message ID you want to fetch

url = f'https://discord.com/api/v10/channels/{MY_CHANNEL_ID}/messages/{MY_MESSAGE_ID}'
result = requests.get(
    url, 
    headers={
        'Authorization': f'Bot {BOT_TOKEN}',
        'Content-Type': 'application/json',
    }, 
)
print(result)
```

### Why are these bad?

#### First Example
This is monkey patching the discord.py library to make an HTTP request from the library work. This also utilizes `asyncio.run` **twice** which can be time consuming and slow.

#### Second Example
This example, although better than the first, will still be unnecessary due to the nature of it. I personally can't think of a context in which you would need to use the Discord API without a Discord bot present - ignoring webhooks of course. Maybe you could elaborate? 

## Alternatives
I encourage you to think about *why* you need to fetch messages from the Discord API in a synchronous context. Could there be a better solution than Discord's API? Could you use a database to assist you and limit API calls? Have you done research for your use case to see if there's an alternative solution?

It's a bit difficult for me to provide you an exact solution, as I have very little information. Feel free to elaborate on your situation and I can provide some more clarity for you if needed :)

**Reply by adrian141b** [reply]
Thanks, but i have a problem with authorization, i received an errror:
'{"message": "401: Unauthorized", "code": 0}'
i checked the token and url and there are the valid. 

I have another system that handles the time in which messages are downloaded after downloading messages, i analyzing the message process them, etc.... which involves me having to finish the code that downloads the messages

**Reply by trevorflahardy** [reply]
> Thanks, but i have a problem with authorization, i received an errror: '{"message": "401: Unauthorized", "code": 0}' i checked the token and url and there are the valid.
> 
> I have another system that handles the time in which messages are downloaded after downloading messages, i analyzing the message process them, etc.... which involves me having to finish the code that downloads the messages

Both the `requests` example I provided and the `hacked discord.py example` (with a bit of tweaking) works for me. Is your bot's token invalid?

### Requests
![image](https://user-images.githubusercontent.com/75498301/179576470-65292375-ef79-4a43-811e-6924c56540dd.png)
![image](https://user-images.githubusercontent.com/75498301/179576497-6975dffc-2283-4fc8-9777-0b2facd8ef28.png)

### Hacked Dpy
![image](https://user-images.githubusercontent.com/75498301/179577834-edb42567-6570-4f88-ae94-9bacaa1c5fbb.png)
![image](https://user-images.githubusercontent.com/75498301/179577851-844d15fd-c8d9-4342-b43a-a480bf57d088.png)

**Reply by crizCraig** [reply]
> **Second Example**
> This example, although better than the first, will still be unnecessary due to the nature of it. I personally can't think of a context in which you would need to use the Discord API without a Discord bot present - ignoring webhooks of course. Maybe you could elaborate?

Seems like the requests example is using a bot, no?

**Reply by trevorflahardy** [reply]
Craig it has been 546 days since this discussion was active, it may be best to leave it in the grave.

While we're here,
>Seems like the requests example is using a bot, no?

The term "without a Discord bot" implies using discord.py without creating a connection to the gateway (ie, without using a bot for its intended purpose), and rather using the library as a standalone API wrapper for the Discord API endpoints. If you have more questions or comments feel free to open a help forum in [the discord.py server](https://discord.gg/dpy) and mention me, my Discord can be found on my Github profile.

Cheers
